### PR TITLE
views: user identity fix

### LINKED ISSUE #63: oauth2server: user identity is not loaded when token is used
**Problem:**
The flask-principal `Identity` is not set to the token's user when a token is used to access HTTP views.

**Suggested solution:**
When token are used, [`invenio_oauth2server.views.server.login_oauth2_user`](https://github.com/inveniosoftware/invenio-oauth2server/blob/8d7c81babb8fe22d8bd3e85b23870edd89eaa9a7/invenio_oauth2server/views/server.py#L53) sets the `current_user` but does not set the flask-principal `Identity`.

This can be fixed by also sending the `identity_changed` signal in `invenio_oauth2server.views.server.login_oauth2_user`.

**Simple test case:**

```python
from __future__ import absolute_import, print_function

import os

from flask import Flask, url_for, request, abort
from flask_login import current_user
from flask_babelex import Babel
from flask_breadcrumbs import Breadcrumbs
from flask_cli import FlaskCLI
from flask_mail import Mail
from flask_menu import Menu
from flask_oauthlib.provider import OAuth2Provider
from invenio_accounts import InvenioAccounts
from invenio_access import InvenioAccess
from invenio_accounts.views import blueprint as accounts_blueprint
from invenio_db import InvenioDB, db

from invenio_oauth2server import InvenioOAuth2Server, current_oauth2server, \
    require_api_auth, require_oauth_scopes, InvenioOAuth2ServerREST
from invenio_oauth2server.models import Scope, Token
from invenio_oauth2server.views import server_blueprint, settings_blueprint
from invenio_oauth2server.provider import oauth2
from flask_security.utils import encrypt_password
from invenio_db import db
from sqlalchemy_utils.functions import create_database, drop_database
from flask_principal import Permission, identity_loaded, identity_changed, \
    Identity

# Create Flask application
app = Flask(__name__)
app.config.update(
    SERVER_NAME='localhost:5000',
    CELERY_ALWAYS_EAGER=True,
    CELERY_CACHE_BACKEND="memory",
    CELERY_EAGER_PROPAGATES_EXCEPTIONS=True,
    CELERY_RESULT_BACKEND="cache",
    OAUTH2SERVER_CACHE_TYPE='simple',
    OAUTHLIB_INSECURE_TRANSPORT=True,
    TESTING=True,
    SECRET_KEY='test_key',
    LOGIN_DISABLED=False,
    SQLALCHEMY_TRACK_MODIFICATIONS=True,
    SQLALCHEMY_DATABASE_URI=os.getenv('SQLALCHEMY_DATABASE_URI',
                                      'sqlite:///example.db'),
)
FlaskCLI(app)
Babel(app)
Mail(app)
Menu(app)
Breadcrumbs(app)
InvenioDB(app)
OAuth2Provider(app)

accounts = InvenioAccounts(app)
app.register_blueprint(accounts_blueprint)
access = InvenioAccess(app)

InvenioOAuth2Server(app)
InvenioOAuth2ServerREST(app)

# register blueprints
app.register_blueprint(settings_blueprint)
app.register_blueprint(server_blueprint)

result = {'loaded': False}

# print on identity change
def on_identity_loaded(sender, identity):
    result['loaded'] = True
    print('IDENTITY LOADED: identity={}'.format(identity))

identity_loaded.connect_via(app)(on_identity_loaded)

@app.route('/', methods=['GET'])
@require_api_auth()
@require_oauth_scopes('test:scope')
def example():
    """Index."""
    # Uncomment this to make the test pass
    # identity_changed.send(app,
    #                       identity=Identity(current_user.id))
    return "hello world"


with app.app_context():
    create_database(db.engine.url)
    db.create_all()
    # Register a test scope
    current_oauth2server.register_scope(Scope('test:scope'))


    creator = accounts.datastore.create_user(
        email='reader@inveniosoftware.org',
        password=encrypt_password('123456'),
        active=True,
    )
    db.session.commit()

    token = Token.create_personal(
        'mytoken', creator.id,
        scopes=['test:scope']
    )
    db.session.commit()

    with app.test_client() as client:
        headers = [('Accept', 'application/json'),
                   ('Authorization',
                    'Bearer {}'.format(token.access_token))
                   ]
        request_res = client.get(
            url_for('example'),
            headers=headers)
        assert request_res.status_code == 200
        assert result['loaded'] == True

```

### PR DESCRIPTION
**Pull request by nharraud**
* FIX Sets user identity when access token is used on the
  REST API (closes #63).

Signed-off-by: Nicolas Harraudeau <nicolas.harraudeau@cern.ch>

### CODE REVIEW

File: tests/test_decorators.py [hunk 1]
```diff
@@ -116,7 +114,7 @@ def test_access_login_required(resource_fixture):
         assert 302 == res.status_code
         assert 'Set-Cookie' in res.headers
         assert res.headers[
-            'Location'].startswith('http://localhost/login/?next=')
+            'Location'].startswith('http://localhost/api/login/?next=')
```
Comment (nharraud) [inline tests/test_decorators.py]: Note that this modification is strange but I did not found any way to make the REST API redirect to the UI login.

File: tests/test_server.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+from invenio_db import db
+
+from invenio_oauth2server.models import Token
+
+
+def test_user_identity_init(app):
+    """Test that user identity is loaded properly when a token is used."""
+    api_app = app.api_app
+    with app.app_context():
+        with db.session.begin_nested():
```
Comment (lnielsen) [inline tests/test_server.py]: Perhaps possible to use existing test fixture?

File: tests/test_server.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            ]
+            user_headers = [
+                ('Accept', 'application/json'),
+                ('Authorization', 'Bearer {}'.format(token.access_token))
+            ]
+        db.session.commit()
+
+        result = {'identity': None}
+
+        @api_app.route('/test_user_identity_init', methods=['GET'])
```
Comment (lnielsen) [inline tests/test_server.py]: Add in resource fixture?

File: tests/conftest.py [hunk 4]
```diff
@@ -260,8 +281,15 @@ def resource_fixture(app):
     """Fixture that contains the test data for models tests."""
     from flask import request
     from invenio_oauth2server.proxies import current_oauth2server
+    api_app = app.api_app
```
Comment (jirikuncar) [inline tests/conftest.py]: Can we use `api_app = app.wsgi_app.mounts['/app']`?

Comment (jirikuncar) [inline tests/conftest.py]: `app_app = app.wsgi_app.mounts['/api'].im_self`

Comment (nharraud) [inline tests/conftest.py]: @jirikuncar This does not work in python 3. I'll use `six.get_method_self` in a dedicated fixture.

File: tests/conftest.py [hunk 5]
```diff
@@ -103,6 +123,12 @@ def teardown():
     return app
 
 
+@pytest.fixture()
+def api_app(app):
+    """Retrieve the REST API application."""
+    return get_method_self(app.wsgi_app.mounts['/api'])
```
Comment (jirikuncar) [inline tests/conftest.py]: 👍

**Pull request merged**